### PR TITLE
feat: Add compareDocumentPosition method from level 3 spec.

### PR DESCRIPTION
**Pull request by zorkow**
The PR adds an implementation of level 3 method [`compareDocumentPosition`](https://dom.spec.whatwg.org/#dom-node-comparedocumentposition). This is an updated version of the code suggested in https://github.com/jindw/xmldom/pull/113

The code ran for many years error-free in [Speech Rule Engine](https://github.com/speech-rule-engine/speech-rule-engine) together with  [wicked-good-xpath](https://github.com/google/wicked-good-xpath).

### CODE REVIEW

File: lib/dom.js [hunk 1]
```diff
@@ -130,6 +130,70 @@ var INVALID_MODIFICATION_ERR = (ExceptionCode.INVALID_MODIFICATION_ERR = ((Excep
 var NAMESPACE_ERR = (ExceptionCode.NAMESPACE_ERR = ((ExceptionMessage[14] = 'Invalid namespace'), 14));
 var INVALID_ACCESS_ERR = (ExceptionCode.INVALID_ACCESS_ERR = ((ExceptionMessage[15] = 'Invalid access'), 15));
 
+// compareDocumentPosition bitmask results
+var DOCUMENT_POSITION_DISCONNECTED = 1;
+var DOCUMENT_POSITION_PRECEDING = 2;
+var DOCUMENT_POSITION_FOLLOWING = 4;
+var DOCUMENT_POSITION_CONTAINS = 8;
+var DOCUMENT_POSITION_CONTAINED_BY = 16;
+var DOCUMENT_POSITION_IMPLEMENTATION_SPECIFIC = 32;
```
Comment (karfau) [inline lib/dom.js:139]: It looks like they should be available on the Node interface, just like the values for `nodeType`:
https://dom.spec.whatwg.org/#ref-for-dom-node-document_position_disconnected
Which should be helpful when adding tests.

**Review by karfau — COMMENTED** [review]
Thank you for picking up this topic and providing an implementation and happy to hear that it was already proven to work for a long time.

I added some comments/remarks/questions. Let me know what you think.

I would also prefer to have tests for this method, to make sure it doesn't break in the future.
Are you able and willing to extend your PR by adding those tests?

**Comment by zorkow** [comment]
Thanks for the review. I'll follow up with revisions after the Easter break.

**Comment by zorkow** [comment]
> 
> I would also prefer to have tests for this method, to make sure it doesn't break in the future. Are you able and willing to extend your PR by adding those tests?

Willing yes, able remains to be seen. But I know what goes wrong in SRE when `compareDocumentPosition` is missing, so we should be able to infer some tests from that.

**Comment by karfau** [comment]
Great. You can push whatever you are able to do and let me know in case I should finalize it.

File: lib/dom.js [hunk 2]
```diff
@@ -603,6 +667,29 @@ Node.prototype = {
 		var prefix = this.lookupPrefix(namespaceURI);
 		return prefix == null;
 	},
+	// Introduced in DOM Level 3:
```
[the remark this answers is not in the document]
Comment (zorkow) [reply lib/dom.js:658]: Added `jsdoc`.

File: lib/dom.js [hunk 3]
```diff
@@ -603,6 +667,29 @@ Node.prototype = {
...
+	compareDocumentPosition: function (o) {
+		if (this === o) return 0;
+		if (this.ownerDocument !== o.ownerDocument)
+			return (
+				DOCUMENT_POSITION_DISCONNECTED |
+				DOCUMENT_POSITION_IMPLEMENTATION_SPECIFIC |
```
[the remark this answers is not in the document]
Comment (zorkow) [reply lib/dom.js]: _Again my disclaimer: this was not originally my code._ 

I suppose as this is a bitmask, `|` seemed a more natural choice. I did some ad hoc benchmarking and in modern `Node` the bit operation is only fractionally faster. So I moved to `+`.

File: lib/dom.js [hunk 4]
```diff
@@ -603,6 +667,29 @@ Node.prototype = {
 		var prefix = this.lookupPrefix(namespaceURI);
 		return prefix == null;
 	},
+	// Introduced in DOM Level 3:
+	compareDocumentPosition: function (o) {
```
[the remark this answers is not in the document]
Comment (zorkow) [reply lib/dom.js]: Done.

File: lib/dom.js [hunk 5]
```diff
@@ -130,6 +130,70 @@ var INVALID_MODIFICATION_ERR = (ExceptionCode.INVALID_MODIFICATION_ERR = ((Excep
...
+ * Computes a new guid.
+ * @return {string} A new guid.
+ */
+function guid() {
+	return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
+		var r = (Math.random() * 16) | 0;
+		var v = c === 'x' ? r : (r & 0x3) | 0x8;
+		return v.toString(16);
+	});
+}
```
[the remark this answers is not in the document]
Comment (zorkow) [reply lib/dom.js]: The spec suggests to use `Math.random()`, which I've done now.

> Should this comparison be consistent across serialization and deserialization or 
> should it not be?

Does not have to be. Deserialization will give you a new object, thus a new DOM. All we have to ensure is that it is consistent over multiple comparison with the same DOMs.

File: lib/dom.js [hunk 6]
```diff
@@ -130,6 +130,70 @@ var INVALID_MODIFICATION_ERR = (ExceptionCode.INVALID_MODIFICATION_ERR = ((Excep
...
+var DOCUMENT_POSITION_CONTAINED_BY = 16;
+var DOCUMENT_POSITION_IMPLEMENTATION_SPECIFIC = 32;
+
+//helper functions for compareDocumentPosition
+/**
+ * Construct a parent chain for a node.
+ * @param {Node} node The start node.
+ * @return {Node[]} The parent chain.
+ */
+//TODO: this can be members
```
[the remark this answers is not in the document]
Comment (zorkow) [reply lib/dom.js:147]: I don't think the comment makes sense. Removed. _Again not my original comment._

File: lib/dom.js [hunk 7]
```diff
@@ -603,6 +667,29 @@ Node.prototype = {
 		var prefix = this.lookupPrefix(namespaceURI);
 		return prefix == null;
 	},
+	// Introduced in DOM Level 3:
+	compareDocumentPosition: function (o) {
+		if (this === o) return 0;
+		if (this.ownerDocument !== o.ownerDocument)
```
[the remark this answers is not in the document]
Comment (zorkow) [reply lib/dom.js]: I've now rewritten the function so it implements the full specification. It follows the spec pretty much to the letter, including the `this` aliasing, which can cause linting problems.

**Review by karfau — APPROVED** [review]
Such a great contribution, thx a lot.
It will be part of version 0.9.0, I will soon release a pre-release version using the `next` label.

**Pull request merged**